src/shared/browserReplay/promotedSkill.ts

```
@@ -0,0 +1,498 @@
+}
+
+/**
+ * Reduce agent-chosen text to a single safe line.
+ *
+ * Used for the flow name inside a hint. Newlines and control characters are
+ * removed, not escaped: the danger is a value that ends the hint line and
+ * starts what looks like a new directive, and a visible `\n` is not that.
+ */
+export function safeHintText(value: unknown, max = MAX_CONTRACT_CHARS): string {
+  if (typeof value !== 'string') return '';
+  return value.replace(/[\u0000-\u001f\u007f]+/g, ' ').replace(/\s+/g, ' ').trim().slice(0, max);
```

ESLint

[error] Unexpected control character(s) in regular expression: \x00, \x1f.

(no-control-regex)